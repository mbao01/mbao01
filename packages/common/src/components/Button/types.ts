import { type ButtonHTMLAttributes } from "react";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "./constants";

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof getButtonClasses> & {
    asChild?: boolean;
  };
