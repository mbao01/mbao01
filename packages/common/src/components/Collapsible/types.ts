import * as CollapsiblePrimitive from "@radix-ui/react-collapsible";

export type CollapsibleProps = React.ComponentPropsWithoutRef<typeof CollapsiblePrimitive.Root>;
