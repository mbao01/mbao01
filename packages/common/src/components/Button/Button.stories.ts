import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    variant: 'default',
    label: 'Button',
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: 'primary',
    label: 'Button',
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    label: 'Wide Button',
  },
};

export const TinyButton: Story = {
  args: {
    size: 'xs',
    label: 'Tiny',
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    label: 'Click me!',
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    label: 'Click me?',
  },
};

export const LoadingButton: Story = {
  args: {
    loading: true,
    label: 'Loading',
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: 'btn-block',
    label: 'Block Button',
  },
};
