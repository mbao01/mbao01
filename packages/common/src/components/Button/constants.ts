import { cva } from "../../libs";

export const getButtonClasses = cva("btn disabled:text-opacity-50", {
  variants: {
    variant: {
      accent: "btn-accent",
      default: "btn-default",
      error: "btn-error",
      ghost: "btn-ghost",
      info: "btn-info",
      link: "btn-link",
      neutral: "btn-neutral",
      primary: "btn-primary",
      secondary: "btn-secondary",
      success: "btn-success",
      warning: "btn-warning",
    },
    outline: {
      true: "btn-outline",
    },
    rounded: {
      xs: "rounded-xs",
      sm: "rounded-sm",
      md: "rounded-md",
      lg: "rounded-lg",
    },
    wide: {
      true: "btn-wide",
    },
    size: {
      xs: "btn-xs",
      sm: "btn-sm",
      md: "btn-md",
      lg: "btn-lg",
    },
    isLoading: {
      true: "relative overflow-hidden",
    },
  },
  compoundVariants: [
    {
      size: undefined,
      className: "min-h-fit h-10",
    },
    {
      rounded: undefined,
      className: "rounded-md",
    },
    {
      variant: undefined,
      outline: true,
      className: "border-neutral-content",
    },
  ],
});
