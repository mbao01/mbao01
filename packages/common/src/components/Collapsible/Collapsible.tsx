"use client";

import * as CollapsiblePrimitive from "@radix-ui/react-collapsible";
import { type CollapsibleProps } from "@radix-ui/react-collapsible";

const Collapsible = (props: CollapsibleProps) => (
  <CollapsiblePrimitive.Root {...props} />
);

Collapsible.Trigger = CollapsiblePrimitive.CollapsibleTrigger;
Collapsible.Content = CollapsiblePrimitive.CollapsibleContent;

export { Collapsible };
