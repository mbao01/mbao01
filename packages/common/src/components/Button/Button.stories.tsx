import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: "Button",
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: "primary",
    children: "Button",
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    children: "Wide Button",
  },
};

export const TinyButton: Story = {
  args: {
    size: "xs",
    children: "Tiny",
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    children: "Click me!",
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    children: "Click me?",
  },
};

export const LoadingButton: Story = {
  args: {
    isLoading: true,
    children: "Loading",
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: "btn-block",
    children: "Block Button",
  },
};

export const SlotButton: Story = {
  args: {
    asChild: true,
    className: "btn-block",
    children: (
      <a href="#" rel="noopener">
        Link
      </a>
    ),
  },
};
