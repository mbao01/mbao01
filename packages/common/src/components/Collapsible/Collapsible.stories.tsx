import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { CaretSortIcon } from "@radix-ui/react-icons";
import { type CollapsibleProps } from "./types";
import { Collapsible } from "./Collapsible";
import { Button } from "../Button";

const withCollapsible = (
  _: StoryFn,
  context: StoryContext<CollapsibleProps>
) => {
  return (
    <Collapsible {...context.args} className="w-[350px] space-y-2">
      <div className="flex items-center justify-between space-x-4 px-4">
        <h4 className="text-sm font-semibold">@mbao01 forked 3 repositories</h4>
        <Collapsible.Trigger asChild>
          <Button variant="ghost" size="sm" className="disabled:bg-transparent">
            <CaretSortIcon className="h-4 w-4" />
            <span className="sr-only">Toggle</span>
          </Button>
        </Collapsible.Trigger>
      </div>
      <div className="rounded-md border px-4 py-2 font-mono text-sm shadow-sm">
        @radix-ui/primitives
      </div>
      <Collapsible.Content className="space-y-2">
        <div className="rounded-md border px-4 py-2 font-mono text-sm shadow-sm">
          @radix-ui/colors
        </div>
        <div className="rounded-md border px-4 py-2 font-mono text-sm shadow-sm">
          @stitches/react
        </div>
      </Collapsible.Content>
    </Collapsible>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Collapsible,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withCollapsible],
} satisfies Meta<typeof Collapsible>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const OpenedByDefault: Story = {
  args: {
    defaultOpen: true,
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
  },
};
