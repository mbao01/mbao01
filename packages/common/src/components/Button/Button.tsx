import React from "react";
import { Slot, Slottable } from "@radix-ui/react-slot";
import { cn } from "../../utilities";
import { Loading } from "../Loading";
import { getButtonClasses } from "./constants";
import { type ButtonProps } from "./types";

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ asChild, className, outline, children, isLoading, variant, size, wide, ...props }, ref) => {
    const Comp = asChild ? Slot : "button";

    return (
      <Comp
        ref={ref}
        className={cn(getButtonClasses({ size, wide, outline, variant, isLoading }), className)}
        {...props}
      >
        <Slottable>{children}</Slottable>
        {isLoading ? (
          <span
            className="absolute left-0 top-0 flex h-full w-full items-center justify-center bg-[inherit]"
            data-testid="loading"
          >
            <Loading />
          </span>
        ) : null}
      </Comp>
    );
  }
);

Button.displayName = "Button";

export { Button };
