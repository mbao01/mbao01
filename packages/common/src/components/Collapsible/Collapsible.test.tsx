import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Collapsible } from "./Collapsible";
import { Button } from "../Button";

describe("Collapsible", () => {
  const renderCollapsible = () => {
    return render(
      <Collapsible>
        <Collapsible.Trigger asChild>
          <Button variant="ghost" size="sm">
            Toggle
          </Button>
        </Collapsible.Trigger>
        <Collapsible.Content>
          <div>@radix-ui/colors</div>
          <div>@stitches/react</div>
        </Collapsible.Content>
      </Collapsible>
    );
  };

  it("renders collapsible trigger with hidden content", () => {
    const { asFragment } = renderCollapsible();

    expect(screen.getByRole("button", { name: "Toggle" })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.queryByText("@radix-ui/colors")).not.toBeInTheDocument();
    expect(screen.queryByText("@stitches/react")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should expand and collapse content", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderCollapsible();
    const toggle = screen.getByRole("button", { name: "Toggle" });
    expect(toggle).toHaveAttribute("data-state", "closed");

    await user.click(toggle);

    expect(toggle).toHaveAttribute("data-state", "open");
    expect(toggle).toHaveAttribute("aria-expanded", "true");
    expect(screen.getByText("@radix-ui/colors")).toBeInTheDocument();
    expect(screen.getByText("@stitches/react")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
