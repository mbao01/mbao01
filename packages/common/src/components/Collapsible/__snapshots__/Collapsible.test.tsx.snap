// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collapsible > renders collapsible trigger with hidden content 1`] = `
<DocumentFragment>
  <div
    data-state="closed"
  >
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      class="btn [&:disabled]:text-opacity-50 btn-ghost btn-sm"
      data-state="closed"
      type="button"
    >
      Toggle
    </button>
    <div
      data-state="closed"
      hidden=""
      id="radix-:r0:"
      style=""
    />
  </div>
</DocumentFragment>
`;

exports[`Collapsible > should expand and collapse content 1`] = `
<DocumentFragment>
  <div
    data-state="open"
  >
    <button
      aria-controls="radix-:r1:"
      aria-expanded="true"
      class="btn [&:disabled]:text-opacity-50 btn-ghost btn-sm"
      data-state="open"
      type="button"
    >
      Toggle
    </button>
    <div
      data-state="open"
      id="radix-:r1:"
      style=""
    >
      <div>
        @radix-ui/colors
      </div>
      <div>
        @stitches/react
      </div>
    </div>
  </div>
</DocumentFragment>
`;
